File: Solution560.java
//560. Subarray Sum Equals K
import java.util.HashMap;
import java.util.Map;

public class Solution560 {
    public static int subarraySum(int[] nums, int k) {
    Map<Integer, Integer> prefixcount = new HashMap<>();
    prefixcount.put(0,1);

    int count = 0;
    int currSum = 0;

    for(int num : nums){
        currSum += num;
        count += prefixcount.getOrDefault(currSum - k, 0);
        prefixcount.merge(currSum, 1, Integer::sum);
    }
    return count;
}

    //Main method
    public static void main(String[] args) {
        // int[] nums = {1,1,1};
        // int k = 2;

        int[] nums = {1,2,3};
        int k = 3;

        System.out.println(subarraySum(nums, k));
    }
}

File: Python/Solution560.py
#560. Subarray Sum Equals K
from typing import List
from collections import Counter

class Solution560:
    def subarraySum(self, nums: List[int], k: int) -> int:
        prefix_sum_count = Counter({0: 1})

        result = 0
        current_sum = 0

        for num in nums:
            current_sum += num

            result += prefix_sum_count[current_sum - k]

            prefix_sum_count[current_sum] += 1

        return result

# nums = [1, 1, 1]
# k = 2

nums = [1,2,3]
k = 3

sol = Solution560()
print(sol.subarraySum(nums, k))
